### ISSUE TITLE
Missing scale and zeropoint for lot of layers on calibrating vision transformer onnx model for INT8 precision

### ISSUE THREAD

**Issue by Shalini194** [NVIDIA/TensorRT#4068]
## Description

 I generated calibration cache for Vision Transformer  onnx model using EntropyCalibration2 method. When trying to generate engine file using cache file for INT8 precision using trtexec, got a lot of missing scale and zeropoint warnings for Constant, Shuffle and SoftMax layers.
 
 [08/08/2024-14:59:41] [I] [TRT] Reading Calibration Cache for calibrator: EntropyCalibration2
[08/08/2024-14:59:41] [I] [TRT] Generated calibration scales using calibration cache. Make sure that calibration cache has latest scales.
[08/08/2024-14:59:41] [I] [TRT] To regenerate calibration cache, please delete the existing one. TensorRT will generate a new calibration cache.
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 7) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 8) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 9) [Shuffle]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 10) [Shuffle]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 20) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 23) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 26) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor /backbone/layers.0/attn/Softmax_output_0, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 31) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[08/08/2024-14:59:41] [W] [TRT] Missing scale and zero-point for tensor (Unnamed Layer* 43) [Constant]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor

I have observed similar warning for SoftMax layer for ResNet18 model. I would like to know the reasons for these warnings. Is int8 support for these layers not present in tensorrt ?

My end goal is to run the entire model with INT8 precision on nvidia hardware. Can you please suggest how can I achieve this ?

## Environment

**TensorRT Version**: 8.6.2.3

**NVIDIA GPU**: Orin Nano 8 GB

**CUDA Version**: 12.2

**CUDNN Version**: 8.9.4

Operating System: Jetpack 6.0

Python Version :  3.10

PyTorch Version : 2.3.0

**Comment by lix19937** [comment]
How do you perform quantification ?   trtexec ptq ?

**Comment by lix19937** [comment]
If you use trtexec -int8 (ptq), softmax run on float type.  BTW, for ViT liked model, it better to use modelopt.

**Comment by Shalini194** [comment]
Thanks for prompt response. 

Used this [calibration script](https://github.com/rmccorm4/tensorrt-utils/tree/master/int8/calibration) to generate calib cache file for ResNet18. Cache file gets generated with the below warning.

![image](https://github.com/user-attachments/assets/e2c2c4d0-334d-4d6e-8a19-82203f108fa4)

Generated the engine with the above cache file using this command:
trtexec --onnx=resnet18.onnx   --saveEngine=resnet18.engine   --int8  --calib=resnet18.cache  --useCudaGraph  --dumpLayerInfo --profilingVerbosity=detailed 

I find that the Gemm and Softmax layers are running in FP32.

![image](https://github.com/user-attachments/assets/e8740415-3b7e-419b-ae7f-80af824e94c1)


I am converting pretrained pytorch model -> ONNX model -> TRT. Is it possible to achieve int8 inference for Gemm and softmax layers using modelopt?

Should we use ONNX PTQ in model optimizer?

**Comment by lix19937** [comment]
> Is it possible to achieve int8 inference for Gemm and softmax layers using modelopt?

Quantization of ['Add', 'AveragePool', 'BatchNormalization', 'Clip', 'Conv', 'ConvTranspose',
    'Gemm', 'GlobalAveragePool', 'MatMul', 'MaxPool', 'Mul'] op types are supported.

Gemm can be run int8, softmax can not.

**Comment by Shalini194** [comment]
@lix19937 , 

For Gemm layer, input datatype is int8 and output datatype is fp32. What will be the **computation precision** in this case ?

Tried to generate tensorrt engines for a couple of classification, segmentation and transformer models. 

The following list of layers always runs in FP32 :

**Concat, Transpose, Reshape, Gather, Resize, ArgMax, LayerNormalization, InstanceNormalization, SoftMax**

And few layers occasionally run in int8  and sometimes fp32 :

**Add, Expand, Pow, MatMul, Erf,** etc

Tried both **implicit quantization** using calibration cache and **explicit quantization** using ModelOpt ONNX PTQ.

I understand that some layers like Reshape, Gather, Concat etc are memory operations and  does not perform any computation & don't need quantization. But, creating tensorrt int8 engine and deploying on hardware doesn't mean purely quantized engine file with all layers running in int8 precision ?

**Comment by lix19937** [comment]
> Thanks for prompt response.
> 
> Used this [calibration script](https://github.com/rmccorm4/tensorrt-utils/tree/master/int8/calibration) to generate calib cache file for ResNet18. Cache file gets generated with the below warning.
> 
> ![image](https://private-user-images.githubusercontent.com/178005810/356944537-e2c2c4d0-334d-4d6e-8a19-82203f108fa4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNTE1MzIsIm5iZiI6MTcyNDE1MTIzMiwicGF0aCI6Ii8xNzgwMDU4MTAvMzU2OTQ0NTM3LWUyYzJjNGQwLTMzNGQtNGQ2ZS04YTE5LTgyMjAzZjEwOGZhNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMFQxMDUzNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yY2M2ZjYzNDg0M2ZkOTgyNzRlZWVhMDM5NzUzZGQ1Nzc3OGJmYjM3YzY2ZTdkNmFjMDYzYjYxNWRjZDJjOTEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.773HEx2z_tr1U1yhT1CqplfkUEBPMNGC8f0VxsS30fU)
> 
> Generated the engine with the above cache file using this command: trtexec --onnx=resnet18.onnx --saveEngine=resnet18.engine --int8 --calib=resnet18.cache --useCudaGraph --dumpLayerInfo --profilingVerbosity=detailed
> 
> I find that the Gemm and Softmax layers are running in FP32.
> 
> ![image](https://private-user-images.githubusercontent.com/178005810/356945231-e8740415-3b7e-419b-ae7f-80af824e94c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNTE1MzIsIm5iZiI6MTcyNDE1MTIzMiwicGF0aCI6Ii8xNzgwMDU4MTAvMzU2OTQ1MjMxLWU4NzQwNDE1LTNiN2UtNDE5Yi1hZTdmLTgwYWY4MjRlOTRjMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMFQxMDUzNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YTExMGY2OWM4MTdjNDkxYWM1ZTBkNGJiMjUzMDY1MTQ1Njc2ZjFjNDUxOTAzNTQwMzE1OThiMGU4N2Y2YWE1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.m92kEmySJt6dz5j8NLofbbiwYd-xnab0i4NhK_UwJdg)
> 
> I am converting pretrained pytorch model -> ONNX model -> TRT. Is it possible to achieve int8 inference for Gemm and softmax layers using modelopt?
> 
> Should we use ONNX PTQ in model optimizer?

By pytorch_quantization  get follow:   
![image](https://github.com/user-attachments/assets/b8145527-8fbf-4a2c-8a80-40700e827c7c)

![image](https://github.com/user-attachments/assets/75a2d2cd-8f6b-45dd-8e46-8c1abeb203b6)

**Comment by YixuanSeanZhou** [comment]
@Shalini194 Quick question: when doing modelopt quantization for ResNet18, did you see any regression in accuracy when running the model using TensorRT (comparing to OnnxRT or Pytorch)? 

Were you quantizing the Conv layers? And in the final TRT engine, were the Conv and Batchnorm / Relu being fused together as a single INT8 op? 

Thanks!

**Comment by Shalini194** [comment]
> @Shalini194 Quick question: when doing modelopt quantization for ResNet18, did you see any regression in accuracy when running the model using TensorRT (comparing to OnnxRT or Pytorch)?
> 
> Were you quantizing the Conv layers? And in the final TRT engine, were the Conv and Batchnorm / Relu being fused together as a single INT8 op?
> 
> Thanks!

@YixuanSeanZhou , 

I was able to retain the Top-1 and Top-5 accuracy for the ModelOpt quantized ResNet18 compared to ONNX and PyTorch. 

Yes, all the Conv layers were quantized. In the final TensorRT engine, Conv+BN+ReLU are fused into a single INT8 operation, with all layers running in INT8 except for Softmax. You can refer to the ResNet18 logs I shared in previous comments for more details. Thanks.

**Comment by lix19937** [comment]
> So, can i conclude that due to some reasons (might not be an optimal fit, accuracy issues in int8, reduce quantization noise), few layers such as LayerNormalization, SoftMax, Concat, Reshape, etc are retained in FP32 in order to create an optimal engine ? 

 LayerNormalization, SoftMax usually run in fp32,  shape-type ops(like Concat, Reshape, Permute) follow the previous layer type.

**Comment by Shalini194** [comment]
> > So, can i conclude that due to some reasons (might not be an optimal fit, accuracy issues in int8, reduce quantization noise), few layers such as LayerNormalization, SoftMax, Concat, Reshape, etc are retained in FP32 in order to create an optimal engine ?
> 
> LayerNormalization, SoftMax usually run in fp32, shape-type ops(like Concat, Reshape, Permute) follow the previous layer type.

@lix19937, Thanks again for clarifying. For vision transformer-like models using ModelOptimizer, all the layers after the first Conv layer are mapped as a single layer with the layer type 'Myelin.' The rest of the graph contains multiple MatMul and Add layers that were quantized. How does it run as a single layer with multiple QDQ nodes ?
